framework/base/Security.php

```diff
@@ -493,48 +498,21 @@ public function generateRandomKey($length = 32)
             $this->_randomSource = null;
         }
 
-        // If not on Windows, try reading from /dev/urandom device. If successful, cache
-        // the file pointer in $this->_randomSource.
-        if (is_resource($this->_randomSource)
-            || ($this->_randomSource === null
-                && DIRECTORY_SEPARATOR === '/'
-                && @is_readable('/dev/urandom'))
-        ) {
-            // Either open /dev/urandom or get the cached file pointer.
-            if (is_resource($this->_randomSource)) {
-                $urandomFile = $this->_randomSource;
-            } else {
-                $urandomFile = fopen('/dev/urandom', 'rb');
-                if ($urandomFile) {
-                    // Check the file's inode protection mode is 'character special'.
-                    // NOTE: octal integer literals!
-                    $fstat = fstat($urandomFile);
-                    if (($fstat['mode'] & 0170000) !== 020000) {
-                        fclose($urandomFile);
-                        $urandomFile = null;
-                    }
-                } else {
-                    $urandomFile = null;
-                }
-            }
+        // If not on Windows, test for a /dev/urandom device.
+        if ($this->_randomSource === null && DIRECTORY_SEPARATOR === '/') {
+            // Check it for speacial character device protection mode.
+            $lstat = @lstat(self::DEV_URANDOM);
+            $urandomDevice = $lstat !== false && ($lstat['mode'] & 0170000) === 020000;
```

Comment:
have you evaluated [stat](https://secure.php.net/manual/en/function.stat.php) vs. [lstat](https://secure.php.net/manual/en/function.lstat.php)?

>  lstat() is identical to stat() except it would instead be based off the symlinks status.